DRAFT: Opentelemetery workshop

ChrisKujawa opened the pull request:
## Description

Hey @nicpuppa @remcowesterhoud lets try to clean this up and that we have something mergable by end of the week.

I think we should try to have something least disrupting, and splitting this into two parts client and gateway.

1. For the gateway we properly just need the dependencies to be added, the rest is done by configurations? Per default we set it already to none afaik?
2. For the client we might need to check how we can do this properly or how other libs doing it
    3. Fex providing some default configs when enabled 
    4. Or allowing to override the tracer 

Happy for your input and collaboration on this. I'm likely not able to fully focus on this this week, but might be able to give some inputs.

remcowesterhoud committed:
feat: start implementation of distributed tracing in clients and gateway

nicpuppa committed:
refactor: cleanup useless spans in the httpclient

nicpuppa committed:
feat: remove configuration class from camunda client

Camunda Java client is just a library and doesn't need to know anything about the opentelemetry configuration.
The customer should provide their own opentelemetry configuration, if they don't provide a configuration the client will
just rely on the NoopOpenTelemetry

nicpuppa committed:
refactor: removed unused configuration class

nicpuppa committed:
refactor: change tracer name and add client version

nicpuppa committed:
refactor: remove spans

nicpuppa committed:
feat: add grpc tracing

nicpuppa committed:
refactor: apply spotless

nicpuppa committed:
refactor: remove unused env var

nicpuppa committed:
test: add default opentelemetry configuration

nicpuppa committed:
build: remove unused dependencies

nicpuppa committed:
build: add dependency to the spring client

nicpuppa committed:
test: add default opentelemetry configuration

nicpuppa committed:
refactor: remove custom grpc interceptor implementation

to achieve the same result we can use the implementation available in the instrumentation dependency

nicpuppa committed:
build: remove unused dependency

nicpuppa committed:
test: add default opentelemetry configuration

nicpuppa committed:
refactor: cleanup

remcowesterhoud commented on parent/pom.xml:
```diff
@@ -277,6 +277,45 @@
   <dependencyManagement>
     <dependencies>
+
+      <dependency>
+        <groupId>com.squareup.okhttp3</groupId>
+        <artifactId>okhttp</artifactId>
+        <version>4.11.0</version>
+      </dependency>
+
+      <dependency>
+        <groupId>io.opentelemetry.instrumentation</groupId>
+        <artifactId>opentelemetry-zipkin-spring-boot-starter</artifactId>
+        <version>2.20.1-alpha</version>
+      </dependency>
+
+      <dependency>
+        <groupId>io.opentelemetry.instrumentation</groupId>
+        <artifactId>opentelemetry-grpc-1.6</artifactId>
+        <version>2.20.1-alpha</version>
+      </dependency>
+
+      <dependency>
+        <groupId>com.google.cloud.opentelemetry</groupId>
+        <artifactId>detector-resources-support</artifactId>
+        <version>0.36.0</version>
+      </dependency>
+
+      <dependency>
+        <groupId>io.opentelemetry.contrib</groupId>
+        <artifactId>opentelemetry-gcp-resources</artifactId>
+        <version>1.49.0-alpha</version>
+      </dependency>
+
+      <dependency>
+        <groupId>io.opentelemetry.instrumentation</groupId>
+        <artifactId>opentelemetry-instrumentation-bom</artifactId>
+        <version>2.20.1</version>
+        <type>pom</type>
+        <scope>import</scope>
+      </dependency>
```
🔧 Let's figure out which dependencies we actually need to keep. Fex, zipkin should be obsolete by now.

We also included some alpha dependencies which is generally not a great idea.

remcowesterhoud replied:
Cleaned up in https://github.com/camunda/camunda/pull/38884/commits/de48145d595380c2379f3e53135e1898d24f741d

The grpc alpha dependency is in use. We should find an alternative so we don't use an alpha version.

remcowesterhoud commented on qa/acceptance-tests/src/test/java/io/camunda/it/TemporaryTest.java:
```diff
@@ -0,0 +1,116 @@
+import org.junit.jupiter.api.Test;
+
+@Disabled
+public class TemporaryTest {
```
❌ We need a proper test to verify `OpenTelemetry`

remcowesterhoud commented on zeebe/gateway-rest/pom.xml:
```diff
@@ -24,6 +24,11 @@
   </properties>
 
   <dependencies>
+    <!-- https://mvnrepository.com/artifact/io.opentelemetry.instrumentation/opentelemetry-spring-boot-starter -->
+    <dependency>
+      <groupId>io.opentelemetry.instrumentation</groupId>
+      <artifactId>opentelemetry-spring-boot-starter</artifactId>
+    </dependency>
```
❌ This is not defined in our parent pom.

remcowesterhoud replied:
Nvm it is defined through the bom.

nicpuppa committed:
feat: the spring client should be able to pickup the global opentelemetry configuration

remcowesterhoud committed:
test: delete temporary test

nicpuppa commented on clients/camunda-spring-boot-starter/src/main/java/io/camunda/client/spring/configuration/SpringCamundaClientConfiguration.java:
```diff
@@ -185,6 +186,11 @@ public boolean preferRestOverGrpc() {
 
+  @Override
+  public OpenTelemetry getOpenTelemetry() {
+    throw new UnsupportedOperationException("OpenTelemetry is not supported yet.");
```
we have to support this also in the spring sdk 👍

remcowesterhoud committed:
build: cleanup opentelemetry dependencies

nicpuppa committed:
refactor: apply spotless

nicpuppa committed:
feat: re-add custom grpc interceptor in order to remove alpha dependency

remcowesterhoud committed:
feat: disable opentelemetry in the gateway by default.

nicpuppa committed:
fix: adjust ci

ChrisKujawa reviewed (commented):
Can we get this ready @remcowesterhoud @nicpuppa wdyt? Looks already quite clean, I guesss only smaller sharp edges?

remcowesterhoud commented on clients/camunda-spring-boot-starter/src/main/java/io/camunda/client/spring/configuration/SpringCamundaClientConfiguration.java:
```diff
@@ -185,50 +185,9 @@ public boolean preferRestOverGrpc() {
-      default ->
-          throw new IllegalStateException(
-              String.format("Unrecognized zeebe protocol '%s'", protocol));
-    };
```
They are private method so I'm going to assume they weren't used and we tossed it because of that.

remcowesterhoud committed:
fix: add OTEL parameter

remcowesterhoud committed:
test: temporary test to start some traces

remcowesterhoud committed:
feat: add annotation processing dependency

This dependency allows us to easily add spans to a trace by annotating a method with the @WithSpan annotation

tanjinyi committed:
attempt to add grpc gateway interception

danielkelemen committed:
feat: add OpenTelemetry context interceptor to enrich span details

remcowesterhoud committed:
demoooooo
